### PR TITLE
feat(LINGUIST-86): Password Reset Backend Development + Email Service + Unit Tests

### PR DESCRIPTION
**Pull request by KardelenCeren**

### CODE REVIEW

File: backend/src/main/java/app/linguistai/bmvp/controller/AccountController.java [hunk 1]
```diff
@@ -71,7 +78,7 @@ public ResponseEntity<Object> changePassword(@Valid @RequestBody QChangePassword
     public ResponseEntity<Object> refreshToken(@RequestHeader(HttpHeaders.AUTHORIZATION) String auth) {
         try {
             RRefreshToken newToken = accountService.refreshToken(auth);
-            return Response.create("new accsess token is created", HttpStatus.OK, newToken);
+            return Response.create("new access token is created", HttpStatus.OK, newToken);
         } catch (Exception e) {
             return Response.create(ExceptionLogger.log(e), HttpStatus.BAD_REQUEST);
```
Comment (selimcanglr) [inline backend/src/main/java/app/linguistai/bmvp/controller/AccountController.java:83]: Keske exceptionlarimizdan gusel gusel error mesajlari donse :(

Comment (KardelenCeren) [reply backend/src/main/java/app/linguistai/bmvp/controller/AccountController.java:83]: Complete exception overhaul in another PR

File: backend/src/main/java/app/linguistai/bmvp/model/ResetToken.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.JoinColumn;
+import jakarta.persistence.ManyToOne;
+import jakarta.persistence.Table;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import org.apache.commons.lang3.RandomStringUtils;
+
+
```
Comment (can-aslan) [inline backend/src/main/java/app/linguistai/bmvp/model/ResetToken.java]: duplicate empty line

File: backend/src/main/java/app/linguistai/bmvp/service/AccountService.java [hunk 3]
```diff
@@ -157,4 +163,68 @@ private String encodePassword(String plainPassword) {
...
+
+    public boolean validateResetCode(String email, String resetCode, boolean invalidate) throws Exception{
+        try {
+            User user = accountRepository.findUserByEmail(email).orElse(null);
+            if (user == null) {
+                throw new NotFoundException("User with email [" + email + "] not found");
+            }
+
+            ResetToken resetToken = resetTokenRepository.findByUserAndResetCode(user, resetCode).orElse(null);
+            if(resetToken == null) {
```
Comment (can-aslan) [inline backend/src/main/java/app/linguistai/bmvp/service/AccountService.java]: "if(resetToken == null) {" -> "if (resetToken == null) {"

File: backend/src/main/java/app/linguistai/bmvp/service/AccountService.java [hunk 4]
```diff
@@ -157,4 +163,68 @@ private String encodePassword(String plainPassword) {
...
+            // create a new reset token
+            ResetToken resetToken = new ResetToken(user);
+            return resetTokenRepository.save(resetToken);
+        } catch (Exception e) {
+            System.out.println("Email token generation exception for email");
+            throw e;
+        }
+    }
+
+    public boolean validateResetCode(String email, String resetCode, boolean invalidate) throws Exception{
```
Comment (can-aslan) [inline backend/src/main/java/app/linguistai/bmvp/service/AccountService.java]: "throws Exception{" -> "throws Exception {"

File: backend/src/main/java/app/linguistai/bmvp/service/AccountService.java [hunk 5]
```diff
@@ -157,4 +163,68 @@ private String encodePassword(String plainPassword) {
...
+            System.out.println("Password reset token validation exception");
+            throw e;
+        }
+    }
+
+    private boolean isResetTokenValid(ResetToken resetToken) {
+        return resetToken.getValidUntil() != null && !resetToken.getValidUntil().before(new Date()) && !resetToken.isUsed();
+    }
+
+    public boolean setPassword(String email, String password) throws Exception{
```
Comment (can-aslan) [inline backend/src/main/java/app/linguistai/bmvp/service/AccountService.java]: "throws Exception{" -> "throws Exception {"

File: backend/src/main/java/app/linguistai/bmvp/service/EmailService.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+			String htmlText = templateEngine.process("forgot-password", context);
+			String subject = "Password reset request";
+			sendEmail(recipientEmail, subject, htmlText, image, LOGO_ATTACHMENT_FILE_NAME);
+		} catch (Exception e) {
+			System.out.println("ERROR: Could not send password reset email.");
+			throw e;
+		}
+	}
+
+	private MultipartFile convertImage(String imagePath) throws Exception{
```
Comment (can-aslan) [inline backend/src/main/java/app/linguistai/bmvp/service/EmailService.java]: "throws Exception{" -> "throws Exception {"

File: backend/src/main/java/app/linguistai/bmvp/service/EmailService.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+		try {
+			Resource resource = new ClassPathResource(imagePath);
+			byte[] imageBytes = resource.getInputStream().readAllBytes();
+			String[] parts = imagePath.split("\\.");
+			String extension;
+			if (parts.length > 1) {
+				extension = parts[parts.length - 1].toLowerCase();
+			}
+			else{
```
Comment (can-aslan) [inline backend/src/main/java/app/linguistai/bmvp/service/EmailService.java]: "else{" -> "else {"

**Review by can-aslan — DISMISSED** [review]
LGTM, thank you for the nice unit tests, just left a few formatting comments. After resolving them, we can merge the branch 🚀

**Review by can-aslan — APPROVED** [review]
image

**Pull request merged**